experimental/virtio/src/vsock/socket.rs

```diff
@@ -0,0 +1,364 @@
+            anyhow::bail!("Peer's stream buffer is full.");
+        }
+
+        self.sent_bytes += data_len;
+        let mut packet = Packet::new_data(data, self.config.port, self.config.host_port)?;
+        self.set_credit_info(&mut packet);
+        self.config.vsock.write_packet(&mut packet);
+        Ok(())
+    }
+
+    /// Updates the credit info on a packet to facilitate flow-control.
+    fn set_credit_info(&mut self, packet: &mut Packet) {
+        packet.set_buf_alloc(STREAM_BUFFER_LENGTH.0);
+        packet.set_fwd_cnt(self.processed_bytes.0);
+        self.previous_processed_bytes = self.processed_bytes;
+    }
+
+    /// Reads the payload of the next available data packet, if any are available.
+    fn read_data(&mut self) -> Option<VecDeque<u8>> {
+        // For now we panic if we are disconnected.
+        assert!(self.connection_state == ConnectionState::Connected);
+        let src_port = self.config.host_port;
+        let dst_port = self.config.port;
+        loop {
+            let packet = self.config.vsock.read_filtered_packet(
+                |packet| packet.get_dst_port() == dst_port && packet.get_src_port() == src_port,
+                true,
+            )?;
+            self.peer_buffer_size = Wrapping(packet.get_buf_alloc());
+            self.peer_processed_bytes = Wrapping(packet.get_fwd_cnt());
+            // For now we panic if we receive an invalid op.
+            match packet.get_op().expect("Invalid packet received on stream.") {
+                VSockOp::CreditRequest => {
+                    self.send_control_packet(VSockOp::CreditUpdate).unwrap();
+                }
+                VSockOp::CreditUpdate => {
+                    // We already updated our flow-control tracking data, so do nothing.
+                }
+                VSockOp::Request | VSockOp::Response => {
+                    // For now we panic if we receive an invalid op.
+                    panic!("Invalid packet received on stream.");
+                }
+                VSockOp::Rst => {
+                    self.connection_state = ConnectionState::Disconnected;
+                    return None;
+                }
+                VSockOp::Shutdown => {
+                    self.send_control_packet(VSockOp::Rst).unwrap();
+                    self.connection_state = ConnectionState::Disconnected;
+                    return None;
+                }
+                VSockOp::Rw => {
+                    let data = packet.get_payload();
+                    // TODO(#2876): Avoid copying the buffer slice if possible.
+                    let mut result = VecDeque::with_capacity(data.len());
+                    result.extend(data);
+                    return Some(result);
+                }
+            }
+        }
+    }
+
+    /// Tries once to fill the destination with as much data as is currently available, either in
+    /// the pending buffer or from the next available data packet.
```

Comment:
Could you add one sentence to the comment saying that it returns the number of bytes read.

Reply:
Thanks, good point.